clippy_utils/src/visitors.rs

```diff
@@ -436,3 +436,61 @@ pub fn for_each_value_source<'tcx, B>(
         _ => f(e),
     }
 }
+
+/// Runs the given function for each path expression referencing the given local which occur after
+/// the given expression.
+pub fn for_each_local_use_after_expr<'tcx, B>(
+    cx: &LateContext<'tcx>,
+    local_id: HirId,
+    expr_id: HirId,
+    f: impl FnMut(&'tcx Expr<'tcx>) -> ControlFlow<B>,
+) -> ControlFlow<B> {
+    struct V<'cx, 'tcx, F, B> {
+        cx: &'cx LateContext<'tcx>,
+        local_id: HirId,
+        expr_id: HirId,
+        found: bool,
+        res: ControlFlow<B>,
+        f: F,
+    }
+    impl<'cx, 'tcx, F: FnMut(&'tcx Expr<'tcx>) -> ControlFlow<B>, B> Visitor<'tcx> for V<'cx, 'tcx, F, B> {
+        type NestedFilter = nested_filter::OnlyBodies;
+        fn nested_visit_map(&mut self) -> Self::Map {
+            self.cx.tcx.hir()
+        }
+
+        fn visit_expr(&mut self, e: &'tcx Expr<'tcx>) {
+            if !self.found {
+                if e.hir_id == self.expr_id {
+                    self.found = true;
+                } else {
+                    walk_expr(self, e);
+                }
+                return;
+            }
+            if self.res.is_break() {
+                return;
+            }
+            if path_to_local_id(e, self.local_id) {
+                self.res = (self.f)(e);
+            } else {
+                walk_expr(self, e);
+            }
+        }
+    }
+
+    if let Some(b) = get_enclosing_block(cx, local_id) {
```

Comment:
NIT: Only getting the enclosing body can be a bit problematic, if the expression passed as `expr_id` is inside a nested block like a function. For the current use case and probably most others, this is fine, though. Could you add a comment to the function explaining this?

```rs
let mut v = vec![];
for i in 0..10 {
    v.push(i); // Called for this expression
    println!("{v:?}");
}

v.push(99); // Misses this expression
```

Otherwise, really nice utility! I was thinking of something like this, and might "steal" this implementation. Thank you!

Reply:
What's wrong with getting the enclosing block of the local? In you example it would get the block enclosing `let mut v = vec![];`.

Reply:
On a side note `get_enclosing_block` doesn't work with trait functions. I'll add a fix for that in a bit.

Reply:
> What's wrong with getting the enclosing block of the local?

Nothing, it's just not clear, that this will always be the block of local, it could also be of some random expression. Using the outer block of the local would be ideal IMO :upside_down_face:

Reply:
Ohhh, I misread the code, sorry, this is fine. Thank you!